src/lib.rs

```diff
@@ -36,5 +38,361 @@ pub fn y_py(_py: Python, m: &PyModule) -> PyResult<()> {
+                        {
+                            "__type": "listitem",
+                            "__format": 0,
+                            "__indent": 0,
+                            "__dir": null,
+                            "__value": 2,
+                            "children": [
+                                {
+                                    "__type": "list",
+                                    "__format": 0,
+                                    "__indent": 0,
+                                    "__dir": "ltr",
+                                    "__listType": "number",
+                                    "__tag": "ol",
+                                    "__start": 1,
+                                    "children": [
+                                        {
+                                            "__type": "listitem",
+                                            "__format": 0,
+                                            "__indent": 0,
+                                            "__dir": "ltr",
+                                            "__value": 1,
+                                            "children": [
+                                                {
+                                                    "__type": "text",
+                                                    "__format": 0,
+                                                    "__style": "",
+                                                    "__mode": 0,
+                                                    "__detail": 0,
+                                                    "text": "d"
+                                                }
+                                            ]
+                                        }
+                                    ]
+                                }
+                            ]
+                        }
+                    ]
+                },
+                {
+                    "__type": "paragraph",
+                    "__format": 0,
+                    "__indent": 0,
+                    "__dir": "ltr",
+                    "children": [
+                        {
+                            "__type": "text",
+                            "__format": 0,
+                            "__style": "",
+                            "__mode": 0,
+                            "__detail": 0,
+                            "text": "e"
+                        }
+                    ]
+                }
+            ]
+        });
+
+        let mut result: HashMap<String, Any> = HashMap::new();
+        let doc: Doc = Doc::new();
+        let root: XmlFragmentRef = doc.get_or_insert_xml_fragment("root");
+        doc.transact_mut()
+            .apply_update(Update::decode_v1(RAW_LEXICAL_STATE_AS_UPDATE).unwrap());
+        let txn: Transaction = doc.transact();
+
+        // Update attributes of the root XmlNode
+        let root_map: MapRef = root.clone().into();
```

Comment:
In the python version of this function, this should all be handled within the function itself. For these sorts of things I often have a function like `process` and then `process_inner` or `process_recursive` or something like that. Then the outer one does some extra stuff for the top level and calls the inner one, and the inner one calls itself and finishes up.

Reply:
I moved all functions to `y_doc.rs`, probably names of these functions can be improved a bit, please review.